calico-vpp-agent/watchers/peers_watcher.go

```diff
@@ -235,40 +256,63 @@ func (w *PeerWatcher) resyncAndCreateWatcher(state map[string]*bgpPeer) error {
 			for ip, asn := range ipAsn {
 				existing, ok := state[ip]
 				if ok {
+					w.log.Debugf("peer(update) neighbor ip=%s for BGPPeer=%s", ip, peer.ObjectMeta.Name)
 					existing.SweepFlag = false
-					if existing.AS != asn {
-						existing.AS = asn
-						err := w.updateBGPPeer(ip, asn)
+					oldSecret := ""
+					if existing.BGPPeerSpec.Password != nil && existing.BGPPeerSpec.Password.SecretKeyRef != nil {
+						oldSecret = existing.BGPPeerSpec.Password.SecretKeyRef.Name
+					}
+					newSecret := ""
+					if peer.Spec.Password != nil && peer.Spec.Password.SecretKeyRef != nil {
+						newSecret = peer.Spec.Password.SecretKeyRef.Name
+					}
+					w.log.Debugf("peer(update) oldSecret=%s newSecret=%s SecretUpdated=%t for BGPPeer=%s", oldSecret, newSecret, existing.SecretUpdated, peer.ObjectMeta.Name)
+					if existing.AS != asn || oldSecret != newSecret || existing.SecretUpdated {
+						err := w.updateBGPPeer(ip, asn, &peer.Spec)
 						if err != nil {
 							return errors.Wrap(err, "error updating BGP peer")
 						}
-					}
-					// Else no change, nothing to do
+						existing.AS = asn
+						existing.BGPPeerSpec = peer.Spec.DeepCopy()
+						existing.SecretUpdated = false
+					} // Else no change, nothing to do
 				} else {
 					// New peer
 					w.log.Infof("peer(add) neighbor ip=%s for BGPPeer=%s", ip, peer.ObjectMeta.Name)
-					state[ip] = &bgpPeer{
-						AS:        asn,
-						SweepFlag: false,
-					}
-					err := w.addBGPPeer(ip, asn)
+					err := w.addBGPPeer(ip, asn, &peer.Spec)
 					if err != nil {
 						return errors.Wrap(err, "error adding BGP peer")
 					}
+					state[ip] = &bgpPeer{
+						AS:            asn,
+						SweepFlag:     false,
+						SecretUpdated: false,
+						BGPPeerSpec:   peer.Spec.DeepCopy(),
+					}
 				}
 			}
 		}
 		// Remove all peers that still have sweepflag to true
 		for ip, peer := range state {
 			if peer.SweepFlag {
 				w.log.Infof("peer(del) neighbor ip=%s", ip)
-				err := w.deleteBGPPeer(ip)
+				err := w.deleteBGPPeer(ip, peer.BGPPeerSpec)
 				if err != nil {
 					return errors.Wrap(err, "error deleting BGP peer")
 				}
 				delete(state, ip)
 			}
 		}
+		// Clean up any secrets that are no longer referenced by any bgp peers
+		if w.secretWatcher != nil {
```

Comment:
I'm not sure why we need this mark and sweep step. Couldn't we just remove the secret watcher when the parent BGPpeer is removed ? That probably requires a RefCnt in case multiple BGPPeers watch the same secret, but that'll